# CGAP Annotation Server 


## Data Source

### Population Variant Data
1. **1000 Genome Project**
 * [ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/)
1. **gnomAD**
 * [https://gnomad.broadinstitute.org/downloads](https://gnomad.broadinstitute.org/downloads)
 * genome, exome, SV
1. **ExAC**
 * [http://exac.broadinstitute.org/downloads](http://exac.broadinstitute.org/downloads)
1. **ESP6500**
 * [https://esp.gs.washington.edu/drupal/](https://esp.gs.washington.edu/drupal/)
1. **TOPMed**
 * [https://www.nhlbiwgs.org/topmed-data-access-scientific-community](https://www.nhlbiwgs.org/topmed-data-access-scientific-community)
1. **KAVIAR**
 * [http://db.systemsbiology.net/kaviar/](http://db.systemsbiology.net/kaviar/)
1. **dbSNP**
 * [ftp://ftp.ncbi.nih.gov/snp/](ftp://ftp.ncbi.nih.gov/snp/)
1. **UK10K**
 * [https://www.uk10k.org/](https://www.uk10k.org/)

### Pathogenicity Score
1. **PolyPhen2**
 * [http://genetics.bwh.harvard.edu/pph2/dokuwiki/downloads](http://genetics.bwh.harvard.edu/pph2/dokuwiki/downloads)
* **CADD**
 * [https://cadd.gs.washington.edu/download](https://cadd.gs.washington.edu/download)
* **SIFT**
 * [https://sift.bii.a-star.edu.sg/sift4g/public//Homo_sapiens/](https://sift.bii.a-star.edu.sg/sift4g/public//Homo_sapiens/)
* **PROVEAN**
 * [http://provean.jcvi.org/downloads.php](http://provean.jcvi.org/downloads.php)
* **DANN**
 * [https://cbcl.ics.uci.edu/public_data/DANN/data/](https://cbcl.ics.uci.edu/public_data/DANN/data/)
* **GERP++**
 * [http://mendel.stanford.edu/SidowLab/downloads/gerp/](http://mendel.stanford.edu/SidowLab/downloads/gerp/)
* **PhyloP**
 * [ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/phyloP46way/](ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/phyloP46way/)
* **MutTaster**
 * [http://www.mutationtaster.org/](http://www.mutationtaster.org/)
* **MutAssessor**
 * [http://mutationassessor.org/r3/](http://mutationassessor.org/r3/)
* **FATHMM**
 * [http://fathmm.biocompute.org.uk/downloads.html](http://fathmm.biocompute.org.uk/downloads.html)
* **dbNSFP: functional prediction and annotation of all potential non-synonymous single-nucleotide variants**
 * [https://sites.google.com/site/jpopgen/dbNSFP](https://sites.google.com/site/jpopgen/dbNSFP) 

### Variant Database
1. **NCBI ClinVar**
 * update check period: weekly
 * [ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/](ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/)
* **ICGC**
 * [https://docs.icgc.org/download/downloading-data/](https://docs.icgc.org/download/downloading-data/)
* **COSMIC**
 * [https://cancer.sanger.ac.uk/cosmic/download](https://cancer.sanger.ac.uk/cosmic/download)
* **OMIM**
 * [https://www.omim.org/downloads/](https://www.omim.org/downloads/)
* **GWAS CATALOG**
 * update check period: weekly
 * [https://www.ebi.ac.uk/gwas/downloads](https://www.ebi.ac.uk/gwas/downloads)
* **SNPedia**
 * [https://www.snpedia.com/index.php/SNPedia](https://www.snpedia.com/index.php/SNPedia)
* **Orphanet**
 * [http://www.orphadata.org/cgi-bin/index.php](http://www.orphadata.org/cgi-bin/index.php)
* **DECIPHER (DatabasE of genomiC varIation and Phenotype in Humans using Ensembl Resources)**
 * [https://decipher.sanger.ac.uk/about#downloads/data](https://decipher.sanger.ac.uk/about#downloads/data)
* **Precision Medicine Knowledge Base (PMKB)**
 * API: [https://pmkb.weill.cornell.edu/api/index.html](https://pmkb.weill.cornell.edu/api/index.html)
* **Database of genomic Variants (DGV)**
 * [http://dgv.tcag.ca/dgv/app/downloads?ref=GRCh37/hg19](http://dgv.tcag.ca/dgv/app/downloads?ref=GRCh37/hg19)
* **Clinical Interpretations of Variants in Cancer (CIViC)**
 * [https://civicdb.org](https://civicdb.org)
* **Ensembl**
 * [ftp://ftp.ensembl.org/pub/release-96/variation/vcf/homo_sapiens/](ftp://ftp.ensembl.org/pub/release-96/variation/vcf/homo_sapiens/)

### Gene Database
* **HGNC**
 * update check period: weekly
 * [https://www.genenames.org/](https://www.genenames.org/)
* 


### Non-coding Region Anntoation
1. **ENCODE**
 * [https://www.encodeproject.org/](https://www.encodeproject.org/)
* **Contextual Analysis of TF Occupancy (CATO)**
 * [http://www.mauranolab.org/CATO/](http://www.mauranolab.org/CATO/)  
* **FANTOM CAT**
 * [http://fantom.gsc.riken.jp/cat/](http://fantom.gsc.riken.jp/cat/)
* **FunSeq2**
 * [http://funseq2.gersteinlab.org/downloads](http://funseq2.gersteinlab.org/downloads)
* **RegulomeDB**
 * [http://www.regulomedb.org/downloads](http://www.regulomedb.org/downloads)
* **TargetScan**
 * [http://www.targetscan.org/vert_72/](http://www.targetscan.org/vert_72/)
* **miRBase**
 * [http://www.mirbase.org/](http://www.mirbase.org/)


### Ontology and Diseases
1. **Human Phenotype Ontology (HPO)** 
 * update check period: monthly
 * [https://hpo.jax.org/app/](https://hpo.jax.org/app/)
 * [https://hpo.jax.org/app/download/ontology](https://hpo.jax.org/app/download/ontology)
 * [https://hpo.jax.org/app/download/annotation](https://hpo.jax.org/app/download/annotation)
* **Monarch Disease Ontology (MONDO)**
 * [https://www.ebi.ac.uk/ols/ontologies/mondo](https://www.ebi.ac.uk/ols/ontologies/mondo)
* **Medical Subject Heading (MeSH)**
 * [https://www.nlm.nih.gov/databases/download/mesh.html](https://www.nlm.nih.gov/databases/download/mesh.html)
* **International Classification of Diseases (ICD)**
 * [https://www.who.int/classifications/icd/en/](https://www.who.int/classifications/icd/en/)
* **OMIM**
 * [https://www.omim.org/downloads/](https://www.omim.org/downloads/)
* **NIH Genetic Home Reference(GHR)**
 * update check period: weekly
 * [https://ghr.nlm.nih.gov/](https://ghr.nlm.nih.gov/)
* **NCBI Genetic Testing Registry (GTR)**
 * [ftp://ftp.ncbi.nlm.nih.gov/pub/GTR/data/](ftp://ftp.ncbi.nlm.nih.gov/pub/GTR/data/)
* **ClinicalTrials**
 * [https://clinicaltrials.gov/ct2/home](https://clinicaltrials.gov/ct2/home)
* **GeneReviews**
 * [https://www.ncbi.nlm.nih.gov/books/NBK138605/](https://www.ncbi.nlm.nih.gov/books/NBK138605/) 
* **SFARI Gene**
 * update check period: monthly
 * [https://gene.sfari.org/tools/](https://gene.sfari.org/tools/)
* **PanelApp**
 * [https://panelapp.genomicsengland.co.uk/](https://panelapp.genomicsengland.co.uk/)

### Drug/Chemical DB
1. **NCBI PubChem**
 * update check period: weekly
 * [https://pubchem.ncbi.nlm.nih.gov/](https://pubchem.ncbi.nlm.nih.gov/)
* **Clinical Pharmacogenetics Implementation Consortium (CPIC)**
 * [https://cpicpgx.org/genes-drugs/](https://cpicpgx.org/genes-drugs/)
* **PharmGKB**
 * [https://www.pharmgkb.org/downloads](https://www.pharmgkb.org/downloads)

### Pathway and Gene-gene Interaction
1. **KEGG pathway**
 * [https://www.genome.jp/kegg/pathway.html](https://www.genome.jp/kegg/pathway.html)
* **Reactome**
 * [https://reactome.org/download-data](https://reactome.org/download-data)


### Annotation Tools
1. **ANNOVAR**
 * [http://annovar.openbioinformatics.org/en/latest/](http://annovar.openbioinformatics.org/en/latest/)
* **VEP + LOFTEE**
 * [https://useast.ensembl.org/info/docs/tools/vep/index.html](https://useast.ensembl.org/info/docs/tools/vep/index.html)
* **SNPEff**
 * [http://snpeff.sourceforge.net/download.html](http://snpeff.sourceforge.net/download.html)

### Annotation Portal
1. **Varsome**
 * API: [https://api.varsome.com/](https://api.varsome.com/)



RefSeq, CGD,  UniGene, IARC TP53, UniProt variants, UniProt domains