# CGAP Annotation Server 


## Data Source

### Population Variant Data
1. **1000 Genome Project**
	* [ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/)
2. **gnomAD**
	* [https://gnomad.broadinstitute.org/downloads](https://gnomad.broadinstitute.org/downloads)
	* genome, exome, SV
3. **ExAC**
	* [http://exac.broadinstitute.org/downloads](http://exac.broadinstitute.org/downloads)
4. **ESP6500**
	* [https://esp.gs.washington.edu/drupal/](https://esp.gs.washington.edu/drupal/)
5. **TOPMed**
	* [https://www.nhlbiwgs.org/topmed-data-access-scientific-community](https://www.nhlbiwgs.org/topmed-data-access-scientific-community)
6. **KAVIAR**
	* [http://db.systemsbiology.net/kaviar/](http://db.systemsbiology.net/kaviar/)
7. **dbSNP**
	* [ftp://ftp.ncbi.nih.gov/snp/](ftp://ftp.ncbi.nih.gov/snp/)
8. **UK10K**
	* [https://www.uk10k.org/](https://www.uk10k.org/)

### Pathogenicity Score
1. **PolyPhen2**
	* [http://genetics.bwh.harvard.edu/pph2/dokuwiki/downloads](http://genetics.bwh.harvard.edu/pph2/dokuwiki/downloads)
2. **CADD**
	* [https://cadd.gs.washington.edu/download](https://cadd.gs.washington.edu/download)
3. **SIFT**
	* [https://sift.bii.a-star.edu.sg/sift4g/public//Homo_sapiens/](https://sift.bii.a-star.edu.sg/sift4g/public//Homo_sapiens/)
4. **PROVEAN**
	* [http://provean.jcvi.org/downloads.php](http://provean.jcvi.org/downloads.php)
5. **DANN**
	* [https://cbcl.ics.uci.edu/public_data/DANN/data/](https://cbcl.ics.uci.edu/public_data/DANN/data/)
6. **GERP++**
	* [http://mendel.stanford.edu/SidowLab/downloads/gerp/](http://mendel.stanford.edu/SidowLab/downloads/gerp/)
7. **PhyloP**
	* [ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/phyloP46way/](ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/phyloP46way/)
8. **MutTaster**
	* [http://www.mutationtaster.org/](http://www.mutationtaster.org/)
9. **MutAssessor**
	* [http://mutationassessor.org/r3/](http://mutationassessor.org/r3/)
10. **FATHMM**
	* [http://fathmm.biocompute.org.uk/downloads.html](http://fathmm.biocompute.org.uk/downloads.html)
11. **dbNSFP: functional prediction and annotation of all potential non-synonymous single-nucleotide variants**
	* [https://sites.google.com/site/jpopgen/dbNSFP](https://sites.google.com/site/jpopgen/dbNSFP) 

### Variant Database
1. **NCBI ClinVar**
	* update check period: weekly
	* [ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/](ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/)
2. **ICGC**
	* [https://docs.icgc.org/download/downloading-data/](https://docs.icgc.org/download/downloading-data/)
3. **COSMIC**
	* [https://cancer.sanger.ac.uk/cosmic/download](https://cancer.sanger.ac.uk/cosmic/download)
4. **OMIM**
	* [https://www.omim.org/downloads/](https://www.omim.org/downloads/)
5. **GWAS CATALOG**
	* update check period: weekly
	* [https://www.ebi.ac.uk/gwas/downloads](https://www.ebi.ac.uk/gwas/downloads)
6. **SNPedia**
	* [https://www.snpedia.com/index.php/SNPedia](https://www.snpedia.com/index.php/SNPedia)
7. **Orphanet**
	* [http://www.orphadata.org/cgi-bin/index.php](http://www.orphadata.org/cgi-bin/index.php)
8. **DECIPHER (DatabasE of genomiC varIation and Phenotype in Humans using Ensembl Resources)**
	* [https://decipher.sanger.ac.uk/about#downloads/data](https://decipher.sanger.ac.uk/about#downloads/data)
9. **Precision Medicine Knowledge Base (PMKB)**
	* API: [https://pmkb.weill.cornell.edu/api/index.html](https://pmkb.weill.cornell.edu/api/index.html)
10. **Database of genomic Variants (DGV)**
	* [http://dgv.tcag.ca/dgv/app/downloads?ref=GRCh37/hg19](http://dgv.tcag.ca/dgv/app/downloads?ref=GRCh37/hg19)
11. **Clinical Interpretations of Variants in Cancer (CIViC)**
	* [https://civicdb.org](https://civicdb.org)
12. **Ensembl**
	* [ftp://ftp.ensembl.org/pub/release-96/variation/vcf/homo_sapiens/](ftp://ftp.ensembl.org/pub/release-96/variation/vcf/homo_sapiens/)
13. **GeneDx**
	* [https://www.genedx.com/test-catalog/medical-specialty/rare/](https://www.genedx.com/test-catalog/medical-specialty/rare/)

### Gene Database
1. **HGNC**
	* update check period: weekly
	* [https://www.genenames.org/](https://www.genenames.org/)
2. **GENCODE**
	* update check period: monthly
	* [https://www.gencodegenes.org/human/](https://www.gencodegenes.org/human/)
3. **RefSeq**
	* [ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/](ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/)
4. **UniGene**

### Non-coding Region Anntoation
1. **ENCODE**
	* [https://www.encodeproject.org/](https://www.encodeproject.org/)
2. **Contextual Analysis of TF Occupancy (CATO)**
	* [http://www.mauranolab.org/CATO/](http://www.mauranolab.org/CATO/)  
3. **FANTOM CAT**
	* [http://fantom.gsc.riken.jp/cat/](http://fantom.gsc.riken.jp/cat/)
4. **FunSeq2**
	* [http://funseq2.gersteinlab.org/downloads](http://funseq2.gersteinlab.org/downloads)
5. **RegulomeDB**
	* [http://www.regulomedb.org/downloads](http://www.regulomedb.org/downloads)
6. **TargetScan**
	* [http://www.targetscan.org/vert_72/](http://www.targetscan.org/vert_72/)
7. **miRBase**
	* [http://www.mirbase.org/](http://www.mirbase.org/)
8. **GeneHanser** 
	* [https://www.genecards.org/GeneHancer_version_4-4](https://www.genecards.org/GeneHancer_version_4-4)

### Ontology and Diseases
1. **Human Phenotype Ontology (HPO)** 
	* update check period: monthly
	* [https://hpo.jax.org/app/](https://hpo.jax.org/app/)
	* [https://hpo.jax.org/app/download/ontology](https://hpo.jax.org/app/download/ontology)
	* [https://hpo.jax.org/app/download/annotation](https://hpo.jax.org/app/download/annotation)
2. **Monarch Disease Ontology (MONDO)**
	* [https://www.ebi.ac.uk/ols/ontologies/mondo](https://www.ebi.ac.uk/ols/ontologies/mondo)
3. **Medical Subject Heading (MeSH)**
	* [https://www.nlm.nih.gov/databases/download/mesh.html](https://www.nlm.nih.gov/databases/download/mesh.html)
4. **International Classification of Diseases (ICD)**
	* [https://www.who.int/classifications/icd/en/](https://www.who.int/classifications/icd/en/)
5. **OMIM**
	* [https://www.omim.org/downloads/](https://www.omim.org/downloads/)
6. **NIH Genetic Home Reference(GHR)**
	* update check period: weekly
	* [https://ghr.nlm.nih.gov/](https://ghr.nlm.nih.gov/)
7. **NCBI Genetic Testing Registry (GTR)**
	* [ftp://ftp.ncbi.nlm.nih.gov/pub/GTR/data/](ftp://ftp.ncbi.nlm.nih.gov/pub/GTR/data/)
8. **ClinicalTrials**
	* [https://clinicaltrials.gov/ct2/home](https://clinicaltrials.gov/ct2/home)
9. **GeneReviews**
	* [https://www.ncbi.nlm.nih.gov/books/NBK138605/](https://www.ncbi.nlm.nih.gov/books/NBK138605/) 
10. **SFARI Gene**
	* update check period: monthly
	* [https://gene.sfari.org/tools/](https://gene.sfari.org/tools/)
11. **PanelApp**
	* [https://panelapp.genomicsengland.co.uk/](https://panelapp.genomicsengland.co.uk/)

### Drug/Chemical DB
1. **NCBI PubChem**
	* update check period: weekly
	* [https://pubchem.ncbi.nlm.nih.gov/](https://pubchem.ncbi.nlm.nih.gov/)
2. **Clinical Pharmacogenetics Implementation Consortium (CPIC)**
	* [https://cpicpgx.org/genes-drugs/](https://cpicpgx.org/genes-drugs/)
3. **PharmGKB**
	* [https://www.pharmgkb.org/downloads](https://www.pharmgkb.org/downloads)

### Pathway and Gene-gene Interaction
1. **KEGG pathway**
	* [https://www.genome.jp/kegg/pathway.html](https://www.genome.jp/kegg/pathway.html)
2. **Reactome**
	* [https://reactome.org/download-data](https://reactome.org/download-data)


### Annotation Tools
1. **ANNOVAR**
	* [http://annovar.openbioinformatics.org/en/latest/](http://annovar.openbioinformatics.org/en/latest/)
2. **VEP + LOFTEE**
	* [https://useast.ensembl.org/info/docs/tools/vep/index.html](https://useast.ensembl.org/info/docs/tools/vep/index.html)
	* [https://github.com/konradjk/loftee](https://github.com/konradjk/loftee)
3. **SNPEff**
	* [http://snpeff.sourceforge.net/download.html](http://snpeff.sourceforge.net/download.html)

### Annotation Portal
1. **Varsome**
 * API: [https://api.varsome.com/](https://api.varsome.com/)
