#!/usr/bin/env python
# -*- coding: utf-8 -*-
import sys
import filecmp
import json
import test_conf
sys.path.append('..')
from src import mutanno
from src.mutanno.util import file_util
from src.mutanno.util import proc_util
from src.mutanno.reader import TSIReader
from src.mutanno.util.vcf_util import INFOIDX
from src.mutanno.util.struct_util import get_dict_value as dv
from src.mutanno.validate import AnnotVCFValidator, AnnotTSIValidator



def test_tsiloader():
    tsi = "./data/mc_21775.tsi.gz"
    tsireader = TSIReader({'tsi':tsi})
    # print(tsireader.infoheader)
    v1 = tsireader.get_variant()
    print(v1)
    print(v1.annot)

    assert tsireader.tsi == "./data/mc_21775.tsi.gz"
    assert v1.annotheader == {'VEP': ['Allele', 'Consequence', 'IMPACT', 'SYMBOL', 'Gene', 'Feature_type', 'Feature', 'BIOTYPE', 'EXON', 'INTRON', 'HGVSc', 'HGVSp', 'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', 'DISTANCE', 'STRAND', 'FLAGS', 'VARIANT_CLASS', 'SYMBOL_SOURCE', 'HGNC_ID', 'CANONICAL', 'MANE', 'TSL', 'APPRIS', 'CCDS', 'ENSP', 'SWISSPROT', 'TREMBL', 'UNIPARC', 'GENE_PHENO', 'SIFT', 'PolyPhen', 'DOMAINS', 'miRNA', 'HGVS_OFFSET', 'AF', 'AFR_AF', 'AMR_AF', 'EAS_AF', 'EUR_AF', 'SAS_AF', 'AA_AF', 'EA_AF', 'gnomAD_AF', 'gnomAD_AFR_AF', 'gnomAD_AMR_AF', 'gnomAD_ASJ_AF', 'gnomAD_EAS_AF', 'gnomAD_FIN_AF', 'gnomAD_NFE_AF', 'gnomAD_OTH_AF', 'gnomAD_SAS_AF', 'MAX_AF', 'MAX_AF_POPS', 'CLIN_SIG', 'SOMATIC', 'PHENO', 'PUBMED', 'MOTIF_NAME', 'MOTIF_POS', 'HIGH_INF_POS', 'MOTIF_SCORE_CHANGE', 'MaxEntScan_alt', 'MaxEntScan_diff', 'MaxEntScan_ref', 'TSSDistance', 'SpliceRegion', 'Feature_ncbi'], 'GENES': [''], 'CYTOBAND': ['cytoband'], 'UNIPROT_TRANSMEM': ['UNIPROTKB_AC', 'SCORE', 'STRAND', 'THICK_START', 'THICK_END', 'ANNOTATION_COLOR', 'NO_BLOCK', 'BLOCK_SIZE', 'BLOCK_START', 'ANNOT_ID', 'ANNOT_DESC'], 'DBSNP': ['ID'], 'CONSERVATION': ['PHASTCONS100', 'PHASTCONS30', 'PHASTCONS20', 'PHYLOP100', 'PHYLOP30', 'PHYLOP20', 'GERP', 'SIPHY_OMEGA', 'SIPHY_P'], 'PRIMATEAI': ['refAA', 'altAA', 'strand_1pos_0neg', 'trinucleotide_context', 'UCSC_gene', 'ExAC_coverage', 'primateDL_score'], 'CADD': ['RawScore', 'PHRED'], 'SPLICEAI': ['SYMBOL', 'DS_AG', 'DS_AL', 'DS_DG', 'DS_DL', 'DP_AG', 'DP_AL', 'DP_DG', 'DP_DL', 'MAXDS', 'MAXDSNAME'], 'MAX_POP_AF': ['AF_popmax'], 'GNOMAD': ['ID', 'QUAL', 'FILTER', 'AC', 'AN', 'AF', 'non_par', 'lcr', 'variant_type', 'n_alt_alleles', 'ReadPosRankSum', 'MQRankSum', 'RAW_MQ', 'DP', 'MQ_DP', 'VarDP', 'MQ', 'QD', 'FS', 'SB', 'InbreedingCoeff', 'AS_VQSLOD', 'NEGATIVE_TRAIN_SITE', 'POSITIVE_TRAIN_SITE', 'culprit', 'SOR', 'AC_asj_female', 'AN_asj_female', 'AF_asj_female', 'nhomalt_asj_female', 'AC_eas_female', 'AN_eas_female', 'AF_eas_female', 'nhomalt_eas_female', 'AC_afr_male', 'AN_afr_male', 'AF_afr_male', 'nhomalt_afr_male', 'AC_female', 'AN_female', 'AF_female', 'nhomalt_female', 'AC_fin_male', 'AN_fin_male', 'AF_fin_male', 'nhomalt_fin_male', 'AC_oth_female', 'AN_oth_female', 'AF_oth_female', 'nhomalt_oth_female', 'AC_ami', 'AN_ami', 'AF_ami', 'nhomalt_ami', 'AC_oth', 'AN_oth', 'AF_oth', 'nhomalt_oth', 'AC_male', 'AN_male', 'AF_male', 'nhomalt_male', 'AC_ami_female', 'AN_ami_female', 'AF_ami_female', 'nhomalt_ami_female', 'AC_afr', 'AN_afr', 'AF_afr', 'nhomalt_afr', 'AC_eas_male', 'AN_eas_male', 'AF_eas_male', 'nhomalt_eas_male', 'AC_sas', 'AN_sas', 'AF_sas', 'nhomalt_sas', 'AC_nfe_female', 'AN_nfe_female', 'AF_nfe_female', 'nhomalt_nfe_female', 'AC_asj_male', 'AN_asj_male', 'AF_asj_male', 'nhomalt_asj_male', 'AC_raw', 'AN_raw', 'AF_raw', 'nhomalt_raw', 'AC_oth_male', 'AN_oth_male', 'AF_oth_male', 'nhomalt_oth_male', 'AC_nfe_male', 'AN_nfe_male', 'AF_nfe_male', 'nhomalt_nfe_male', 'AC_asj', 'AN_asj', 'AF_asj', 'nhomalt_asj', 'AC_amr_male', 'AN_amr_male', 'AF_amr_male', 'nhomalt_amr_male', 'nhomalt', 'AC_amr_female', 'AN_amr_female', 'AF_amr_female', 'nhomalt_amr_female', 'AC_sas_female', 'AN_sas_female', 'AF_sas_female', 'nhomalt_sas_female', 'AC_fin', 'AN_fin', 'AF_fin', 'nhomalt_fin', 'AC_afr_female', 'AN_afr_female', 'AF_afr_female', 'nhomalt_afr_female', 'AC_sas_male', 'AN_sas_male', 'AF_sas_male', 'nhomalt_sas_male', 'AC_amr', 'AN_amr', 'AF_amr', 'nhomalt_amr', 'AC_nfe', 'AN_nfe', 'AF_nfe', 'nhomalt_nfe', 'AC_eas', 'AN_eas', 'AF_eas', 'nhomalt_eas', 'AC_ami_male', 'AN_ami_male', 'AF_ami_male', 'nhomalt_ami_male', 'AC_fin_female', 'AN_fin_female', 'AF_fin_female', 'nhomalt_fin_female', 'faf95_afr', 'faf99_afr', 'faf95_sas', 'faf99_sas', 'faf95_adj', 'faf99_adj', 'faf95_amr', 'faf99_amr', 'faf95_nfe', 'faf99_nfe', 'faf95_eas', 'faf99_eas'], 'TOPMED': ['ID', 'QUAL', 'FILTER', 'NS', 'AN', 'AC', 'AF', 'Het', 'Hom', 'VRT', 'FORMAT', 'TOPMED'], 'UK10K': ['ID', 'QUAL', 'FILTER', 'DP', 'VQSLOD', 'AC', 'AN', 'AF', 'AC_TWINSUK', 'AN_TWINSUK', 'AF_TWINSUK', 'AC_ALSPAC', 'AN_ALSPAC', 'AF_ALSPAC', 'AF_AFR', 'AF_AMR', 'AF_ASN', 'AF_EUR', 'AF_MAX', 'ESP_MAF', 'CSQ', 'AC_TWINSUK_NODUP', 'AN_TWINSUK_NODUP', 'AF_TWINSUK_NODUP', 'GRCh37'], 'GENOMIC_SUPER_DUPLICATES': ['name', 'bin', 'score', 'strand', 'otherChrom', 'otherStart', 'otherEnd', 'otherSize', 'uid', 'posBasesHit', 'testResult', 'verdict', 'chits', 'ccov', 'alignfile', 'alignL', 'indelN', 'indelS', 'alignB', 'matchB', 'mismatchB', 'transitionsB', 'transversionsB', 'fracMatch', 'fracMatchIndel', 'jcK', 'k2K'], 'SIMPLE_REPEAT': ['name', 'bin', 'period', 'copyNum', 'consensusSize', 'perMatch', 'perIndel', 'score', 'A', 'C', 'G', 'T', 'entropy', 'sequence'], 'RMSK': ['bin', 'swScore', 'milliDiv', 'milliDel', 'milliIns', 'genoName', 'genoStart', 'genoEnd', 'genoLeft', 'strand', 'repName', 'repClass', 'repFamily', 'repStart', 'repEnd', 'repLeft', 'id'], 'NESTED_REPEATS': ['name', 'bin', 'score', 'strand', 'thickStart', 'thickEnd', 'reserved', 'blockCount', 'blockSizes', 'chromStarts', 'blockStrands', 'id', 'repClass', 'repFamily'], 'MICROSATELLITE': ['name', 'bin'], 'CLINVAR': ['VARIATIONID', 'ALLELEID', 'CLNDN', 'CLNDNINCL', 'CLNDISDB', 'CLNDISDBINCL', 'CLNHGVS', 'CLNREVSTAT', 'CLNSIG', 'CLNSIGCONF', 'CLNSIGINCL', 'CLNVC', 'CLNVCSO', 'CLNVI', 'GENEINFO', 'MC', 'ORIGIN', 'SSR'], 'CLINVAR_SUBMISSION': ['VARIATIONID', 'ClinVarAccession', 'Interpretation', 'DateLastEvaluated', 'ReviewStatus', 'Method', 'Condition', 'AlleleOrigin', 'Submitter', 'SubmitterID', 'Citation'], 'COSMIC': ['Gene name', 'Accession Number', 'Gene CDS length', 'HGNC ID', 'Sample name', 'ID_sample', 'ID_tumour', 'Primary site', 'Site subtype 1', 'Site subtype 2', 'Site subtype 3', 'Primary histology', 'Histology subtype 1', 'Histology subtype 2', 'Histology subtype 3', 'Genome-wide screen', 'GENOMIC_MUTATION_ID', 'LEGACY_MUTATION_ID', 'MUTATION_ID', 'Mutation CDS', 'Mutation AA', 'Mutation Description', 'Mutation zygosity', 'LOH', 'GRCh', 'Mutation genome position', 'Mutation strand', 'SNP', 'Resistance Mutation', 'FATHMM prediction', 'FATHMM score', 'Mutation somatic status', 'Pubmed_PMID', 'ID_STUDY', 'Sample Type', 'Tumour origin', 'Age', 'Original gene name'], 'ENSEMBLANNOT': ['strand', 'ensgid', 'gene_version', 'gene_symbol', 'gene_source', 'gene_biotype', 'type', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version']}
    
    

if __name__ == "__main__":
    test_tsiloader()