

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>5. Annotate Variants (annot) &mdash; MutAnno 0.3 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="6. Datasource List" href="datasource_list.html" />
    <link rel="prev" title="4. Make Single Source File (makedata)" href="make_sourcefile.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> MutAnno
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="options.html">2. Optional arguments</a></li>
<li class="toctree-l1"><a class="reference internal" href="download.html">3. Download Data Source (<code class="docutils literal notranslate"><span class="pre">download</span></code>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_sourcefile.html">4. Make Single Source File (<code class="docutils literal notranslate"><span class="pre">makedata</span></code>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">5. Annotate Variants (<code class="docutils literal notranslate"><span class="pre">annot</span></code>)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-file">5.1. Input file</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#vcf-file-vcf">5.1.1. VCF file (<code class="docutils literal notranslate"><span class="pre">-vcf</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-source-file-sourcefile">5.1.2. Data source file (<code class="docutils literal notranslate"><span class="pre">-sourcefile</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-structure-file-ds">5.1.3. Data structure file (<code class="docutils literal notranslate"><span class="pre">-ds</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#output-file">5.2. Output file</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#annotated-output-file-out">5.2.1. Annotated output file (<code class="docutils literal notranslate"><span class="pre">-out</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#out-file-type-outtype">5.2.2. Out file type (<code class="docutils literal notranslate"><span class="pre">-outtype</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#split-multi-allelic-variant-into-single-allelic-variant-split-multi-allelic-variant">5.3. Split multi-allelic variant into single allelic variant (<code class="docutils literal notranslate"><span class="pre">-split_multi_allelic_variant</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#add-additional-information">5.4. Add additional information</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#add-genotype-genoinfo">5.4.1. Add genotype (<code class="docutils literal notranslate"><span class="pre">-genoinfo</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#add-hg19-coordinates-hg19">5.4.2. Add hg19 coordinates (<code class="docutils literal notranslate"><span class="pre">-hg19</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#remove-annotation-tag">5.5. Remove annotation tag</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="datasource_list.html">6. Datasource List</a></li>
<li class="toctree-l1"><a class="reference internal" href="datastructure.html">7. Data Structure File</a></li>
<li class="toctree-l1"><a class="reference internal" href="datastructure_version.html">8. Data Structure File Version</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotated_vcf.html">9. Annotated VCF file format</a></li>
<li class="toctree-l1"><a class="reference internal" href="benchmark.html">10. Benchmark</a></li>
<li class="toctree-l1"><a class="reference internal" href="version_list.html">11. Version History</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MutAnno</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">5. </span>Annotate Variants (<code class="docutils literal notranslate"><span class="pre">annot</span></code>)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/annotate_variants.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="annotate-variants-annot">
<h1><span class="section-number">5. </span>Annotate Variants (<code class="docutils literal notranslate"><span class="pre">annot</span></code>)<a class="headerlink" href="#annotate-variants-annot" title="Permalink to this headline">¶</a></h1>
<p><strong>MutAnno</strong> can read all variants of VCF file (<code class="docutils literal notranslate"><span class="pre">-vcf</span></code>), annotate them with annotation source file (<code class="docutils literal notranslate"><span class="pre">-sourfile</span></code>) and data structure file (<code class="docutils literal notranslate"><span class="pre">-ds</span></code>), and generate annotated VCF file (<code class="docutils literal notranslate"><span class="pre">-out</span></code>). For the variant annotation, the simple command example is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> \
        <span class="o">-</span><span class="n">vcf</span> <span class="nb">input</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">-</span><span class="n">out</span> <span class="n">out_annotated</span><span class="o">.</span><span class="n">vcf</span> \
        <span class="o">-</span><span class="n">ds</span> <span class="n">datastructure</span><span class="o">.</span><span class="n">json</span> \
        <span class="o">-</span><span class="n">sourfile</span> <span class="n">annotation_sourcefile</span><span class="o">.</span><span class="n">tsi</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<div class="section" id="input-file">
<h2><span class="section-number">5.1. </span>Input file<a class="headerlink" href="#input-file" title="Permalink to this headline">¶</a></h2>
<div class="section" id="vcf-file-vcf">
<h3><span class="section-number">5.1.1. </span>VCF file (<code class="docutils literal notranslate"><span class="pre">-vcf</span></code>)<a class="headerlink" href="#vcf-file-vcf" title="Permalink to this headline">¶</a></h3>
<p><strong>MutAnno</strong> can import text-based VCF file(<code class="docutils literal notranslate"><span class="pre">.vcf</span></code>) or gzipped/bgzipped VCF file (<code class="docutils literal notranslate"><span class="pre">.vcf.gz</span></code>) as input file. <strong>MutAnno</strong> cannot support mutliple VCF files, but a single vcf file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">vcf</span> <span class="nb">input</span><span class="o">.</span><span class="n">vcf</span>
<span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">vcf</span> <span class="nb">input</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<div class="section" id="data-source-file-sourcefile">
<h3><span class="section-number">5.1.2. </span>Data source file (<code class="docutils literal notranslate"><span class="pre">-sourcefile</span></code>)<a class="headerlink" href="#data-source-file-sourcefile" title="Permalink to this headline">¶</a></h3>
<p>If you have single source file which is generated by <code class="docutils literal notranslate"><span class="pre">mutanno</span> <span class="pre">makedata</span></code>, you can assign the file with this option. You can see more details for generation of this single source file in <a class="reference external" href="make_sourcefile.html">here</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">sourcefile</span> <span class="n">annotation_sourcefile</span><span class="o">.</span><span class="n">tsi</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>But, if you don’t have the single source file and only have the multiple downloaded source files, you can add the file path in the <cite>data structure`(`</cite>-ds``) file. In this case, you don’t need to use this <code class="docutils literal notranslate"><span class="pre">-sourcefile</span></code> option.</p>
</div>
</div>
<div class="section" id="data-structure-file-ds">
<h3><span class="section-number">5.1.3. </span>Data structure file (<code class="docutils literal notranslate"><span class="pre">-ds</span></code>)<a class="headerlink" href="#data-structure-file-ds" title="Permalink to this headline">¶</a></h3>
<p>The data structure file includes the annotation feature list being extracted from source file(s). You can assign the structure file with <code class="docutils literal notranslate"><span class="pre">-ds</span></code> option and get more information in <a class="reference external" href="datastructure.html">here</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">ds</span> <span class="n">data_structure</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="output-file">
<h2><span class="section-number">5.2. </span>Output file<a class="headerlink" href="#output-file" title="Permalink to this headline">¶</a></h2>
<div class="section" id="annotated-output-file-out">
<h3><span class="section-number">5.2.1. </span>Annotated output file (<code class="docutils literal notranslate"><span class="pre">-out</span></code>)<a class="headerlink" href="#annotated-output-file-out" title="Permalink to this headline">¶</a></h3>
<p>The output file name is assigned with <code class="docutils literal notranslate"><span class="pre">-out</span></code> option.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">out</span> <span class="n">annotated</span><span class="o">.</span><span class="n">vcf</span>
<span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">out</span> <span class="n">annotated</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<div class="section" id="out-file-type-outtype">
<h3><span class="section-number">5.2.2. </span>Out file type (<code class="docutils literal notranslate"><span class="pre">-outtype</span></code>)<a class="headerlink" href="#out-file-type-outtype" title="Permalink to this headline">¶</a></h3>
<p>You can set out file type (<code class="docutils literal notranslate"><span class="pre">vcf</span></code>, <code class="docutils literal notranslate"><span class="pre">json</span></code>, or <code class="docutils literal notranslate"><span class="pre">vcf</span> <span class="pre">json</span></code>) with <code class="docutils literal notranslate"><span class="pre">-outtype</span></code> option. By default, the out file type is <code class="docutils literal notranslate"><span class="pre">vcf</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">outtype</span> <span class="n">vcf</span>
<span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">outtype</span> <span class="n">json</span>
<span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">outtype</span> <span class="n">vcf</span> <span class="n">json</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="split-multi-allelic-variant-into-single-allelic-variant-split-multi-allelic-variant">
<h2><span class="section-number">5.3. </span>Split multi-allelic variant into single allelic variant (<code class="docutils literal notranslate"><span class="pre">-split_multi_allelic_variant</span></code>)<a class="headerlink" href="#split-multi-allelic-variant-into-single-allelic-variant-split-multi-allelic-variant" title="Permalink to this headline">¶</a></h2>
<p>In the VCF file, there is a multi-allelic variant that has two or more alternative alleles as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#CHROM   POS       ID     REF     ALT</span>
<span class="n">chr1</span>     <span class="mi">2376453</span>   <span class="o">.</span>      <span class="n">A</span>       <span class="n">T</span><span class="p">,</span><span class="n">G</span>
</pre></div>
</div>
<p>For the multi-allelic variant, <strong>MutAnno</strong> can add the annotations of those alternative alleles in corresponding INFO field.</p>
<p>However, if you want to keep the annotation of each alternative allele, you can use <code class="docutils literal notranslate"><span class="pre">-single_source_mode</span></code> option. When you are using this option, <strong>MutAnno</strong> can split the multi alternative alleles into two or more variants using several lines. And it add the annotation of each single variant in each line.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#CHROM   POS       ID     REF     ALT</span>
<span class="n">chr1</span>     <span class="mi">2376453</span>   <span class="o">.</span>      <span class="n">A</span>       <span class="n">T</span>
<span class="n">chr1</span>     <span class="mi">2376453</span>   <span class="o">.</span>      <span class="n">A</span>       <span class="n">G</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">single_source_mode</span>
</pre></div>
</div>
</div>
<div class="section" id="add-additional-information">
<h2><span class="section-number">5.4. </span>Add additional information<a class="headerlink" href="#add-additional-information" title="Permalink to this headline">¶</a></h2>
<div class="section" id="add-genotype-genoinfo">
<h3><span class="section-number">5.4.1. </span>Add genotype (<code class="docutils literal notranslate"><span class="pre">-genoinfo</span></code>)<a class="headerlink" href="#add-genotype-genoinfo" title="Permalink to this headline">¶</a></h3>
<p>If you want to add genotype information of each samples in the INFO field, you can use <code class="docutils literal notranslate"><span class="pre">-genoinfo</span></code> option.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">genoinfo</span>
</pre></div>
</div>
<p>This option can add the following information separated by <code class="docutils literal notranslate"><span class="pre">|(pipe)</span></code>:</p>
<ul class="simple">
<li><p>genotype(number format)</p></li>
<li><p>genotype(base format)</p></li>
<li><p>reference read depth/total depth</p></li>
<li><p>sample ID</p></li>
</ul>
<p>And, each genoinfos are separated by <code class="docutils literal notranslate"><span class="pre">,(comma)</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#INFO</span>
<span class="n">SAMPLEGENO</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="o">|</span><span class="n">C</span><span class="o">/</span><span class="n">C</span><span class="o">|</span><span class="mi">0</span><span class="o">/</span><span class="mi">1778</span><span class="o">|</span><span class="n">NA12877_sample</span><span class="p">,</span><span class="mi">0</span><span class="o">/</span><span class="mi">0</span><span class="o">|</span><span class="n">T</span><span class="o">/</span><span class="n">T</span><span class="o">|</span><span class="mi">2890</span><span class="o">/</span><span class="mi">0</span><span class="o">|</span><span class="n">NA12878_sample</span><span class="p">;</span>
</pre></div>
</div>
</div>
<div class="section" id="add-hg19-coordinates-hg19">
<h3><span class="section-number">5.4.2. </span>Add hg19 coordinates (<code class="docutils literal notranslate"><span class="pre">-hg19</span></code>)<a class="headerlink" href="#add-hg19-coordinates-hg19" title="Permalink to this headline">¶</a></h3>
<p>If the reference version is hg38 (or GRCh38) and you want to add coordinates of hg19 for all variants, use <code class="docutils literal notranslate"><span class="pre">-hg19</span></code> option. With this option, <strong>MutAnno</strong> can append hg19 coordinates using <a class="reference external" href="https://pypi.org/project/pyliftover/">pyliftover</a> librarry.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">hg19</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This option needs chain file that can be automatically downloaded from UCSC (<a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenpath/hg38/liftOver/">http://hgdownload.cse.ucsc.edu/goldenpath/hg38/liftOver/</a>). But if you want to use other chain file, you can assign the file with <code class="docutils literal notranslate"><span class="pre">-chain</span></code> option.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">hg19</span> <span class="o">-</span><span class="n">chain</span> <span class="n">hg38ToHg19</span><span class="o">.</span><span class="n">over</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="remove-annotation-tag">
<h2><span class="section-number">5.5. </span>Remove annotation tag<a class="headerlink" href="#remove-annotation-tag" title="Permalink to this headline">¶</a></h2>
<p><strong>MutAnno</strong> can remove annotations (or tags) that have already been written in INFO fields of VCF file. If you want to remove some annotations, you can use <code class="docutils literal notranslate"><span class="pre">-clean_tag</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mutanno</span> <span class="n">annot</span> <span class="o">-</span><span class="n">clean_tag</span> <span class="n">SpliceAI</span> <span class="n">CLINVAR</span> <span class="n">gnomADgenome</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="datasource_list.html" class="btn btn-neutral float-right" title="6. Datasource List" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="make_sourcefile.html" class="btn btn-neutral float-left" title="4. Make Single Source File (makedata)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, CGAP

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>