Search.setIndex({docnames:["annotate_variants","annotated_vcf","benchmark","cgap_example","datasource_list","datastructure","datastructure_version","download","index","installation","make_sourcefile","options","precalculate_annotation","version_list"],envversion:{"sphinx.domains.c":2,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":3,"sphinx.domains.index":1,"sphinx.domains.javascript":2,"sphinx.domains.math":2,"sphinx.domains.python":2,"sphinx.domains.rst":2,"sphinx.domains.std":1,"sphinx.ext.intersphinx":1,"sphinx.ext.todo":2,"sphinx.ext.viewcode":1,sphinx:56},filenames:["annotate_variants.rst","annotated_vcf.rst","benchmark.rst","cgap_example.rst","datasource_list.rst","datastructure.rst","datastructure_version.rst","download.rst","index.rst","installation.rst","make_sourcefile.rst","options.rst","precalculate_annotation.rst","version_list.rst"],objects:{},objnames:{},objtypes:{},terms:{"00to99":1,"1000gp":1,"3a14825":1,"3a31276":1,"3a65565":1,"4_200614":3,"6ds":3,"874e":1,"boolean":5,"case":[0,1],"class":11,"default":[0,5,7,10,13],"function":[1,5,13],"import":0,"new":1,"true":[5,13],And:0,But:[0,10],For:[0,1],POS:[0,1],The:[0,1,5,10],These:1,With:[0,10],YES:1,abl:1,abov:1,access:1,actual:1,add:[5,6,8,11,13],add_genoinfo:13,added:[],addit:[1,8],after:[1,10],again:1,all:[0,1,7,13],allel:[1,8,11,13],alloc:13,allow:1,alreadi:0,also:1,alt:[0,1],altern:[0,9],among:10,ani:[1,11],annot3tool:12,annot:[5,7,8,13],annotation_sourc:10,annotation_sourcefil:0,annovar:2,anntoat:10,append:0,appli:[1,13],appropri:1,argument:8,ariant_class:11,arrai:1,assign:[0,5,10],assum:1,attribut:8,automat:[0,5,9],avil:5,base:[0,8],basic:[8,10],becaus:9,bed:5,been:0,begin:1,being:0,below:1,benchmark:8,bgm:[3,9],bgzip:0,biallel:13,bin:12,bio:12,blank:[1,13],block:[1,11,13],blocksiz:10,bold:1,both:1,cache_vers:12,cadd:8,calcul:11,can:[0,1,5,7,9,10],cannot:[0,1],cast:1,ccds30547:1,cdcsd:[],cgap:8,chain:[0,3,11,13],chang:13,chr1:[0,1,10],chrom:[0,1],chromosom:13,clean_tag:[0,3,13],clinvar:[0,3,8],clinvar_vcf_fil:4,clone:9,code:8,coding_gene_main_chrom:[3,8,13],collaps:1,com:[3,9],comma:[0,1],command:[0,5,7,8,9,11,13],comparison:2,conf:13,configr:13,consid:1,consist:1,contain:10,content:[],convchrom:10,convers:1,convert:[1,8,13],coordin:[8,11],correspond:[0,1],count:6,creat:1,cse:0,culprit:1,data1:12,data:[1,3,8,13],data_structur:[0,10],databas:8,datasourc:[8,13],datasource_directori:7,datastructur:0,datastructure_fullannot_v0:3,datastructure_gene_v0:3,datastructure_microannot_v0:3,date:[1,4,5],dbmi:[3,9,12],dbnsfp:13,dbnsfptranscript:13,dbsnp:[8,10],debug:[11,13],delimit:5,delin:1,delta:6,depth:0,desc:5,describ:1,descript:[1,5],detail:[0,1],determin:1,dictionari:1,differ:1,direct:1,directli:1,directori:5,doc:4,don:[0,9,10],done:1,download:[0,5,8,10,13],downstream_gene_vari:1,dropbox:3,each:[0,1,10],ebug:11,edit:5,edu:0,effect:1,effici:10,either:1,elasticsearch:1,element:1,embed:1,encapsul:1,encod:13,end:1,enet:11,enoinfo:11,ensembl:12,ensg00000186092:1,ensg00000240361:1,ensp00000334393:1,enst00000335137:1,enst00000492842:1,entri:1,equal:1,etc:1,even:1,exampl:[0,8],excesshet:1,exit:11,expect:1,extern:13,extract:0,fals:5,fasta:12,featur:0,field1:5,field:[0,1,8,11,13],file:[3,7,8,13],filter:[1,13],first:[1,13],fix:13,fixpl:13,follow:[0,1,9],format:[0,5,8,10,13],forth:1,from:[0,6,8,10,13],full:[8,13],fullanno:3,fullannot_source_fil:3,further:1,g19:11,gene:[8,11],gene_id:1,gene_main_chrom:13,gene_nam:1,gener:[0,5,8],geno_info:13,genoinfo:[3,8,13],genom:8,genotyp:[8,11],get:[0,1,13],git:9,github:[3,8],given:1,gnomad:[6,8],gnomadgenom:[0,3],goldenpath:0,gov:4,grab:1,grch38:0,grch38_full_analysis_set_plus_decoy_hla:12,grch38_latest_dbsnp_al:10,grch38d1:12,group:1,gvs:11,gzip:[0,3],hain:11,has:[0,1,13],have:[0,1,10],header:[1,13],heirach:5,help:11,here:[0,10],hg002:[1,13],hg003:13,hg004:13,hg19:[3,7,8,11],hg38:[0,7],hg38tohg19:[0,3],hgdownload:0,hgnc:1,hgv:[11,13],hgvsg:13,histori:8,hms:12,home:12,homo_sapiens_vep:12,homozyg:6,how:1,howev:0,http:[0,3,4,9],ilenc:11,implement:13,includ:[0,10],indic:13,individu:6,info:[0,1,11],inform:[1,8,11],ingle_source_mod:11,input:[3,8,11],insert:[1,10],instal:[8,13],interfac:8,is_avail:5,is_list:5,is_most_severe_transcript:13,item:8,json:[0,3,5,10,11,13],jsonfil:13,keep:0,keep_equal_str:13,kei:1,last:1,lastest:7,later:1,lean_tag:11,level:5,librari:9,librarri:0,liftov:[0,11],line:[0,1,8],list:[0,1,5,8,13],load:11,locat:1,locksiz:11,log:11,loss:1,maestro:3,make:[1,8,13],makedata:[0,3,8,13],manag:10,map:[1,13],mark:1,maxd:6,maximum:6,mean:[1,5],memori:11,merg:[3,13],merge_vep:12,messag:11,metadata:1,metric:11,micro:8,microannot:3,microannot_datasourc:3,minor:13,miss:13,mk446:12,mleac:1,mleaf:1,mode:[11,13],modifi:1,modul:13,more:[0,10],mti:3,multi:[1,8,11,13],multiallel:13,multipl:[0,1,13],must:[1,10],mutanno:[0,1,3,5,7,9,10,11,12,13],mutanno_sourc:7,mutlipl:0,muttano:1,mvp_gene_datasource_v0:3,na12877_sampl:0,na12878_sampl:0,name:[0,1,4,5,7,10,13],ncbi:4,need:[0,7,9],negative_train_sit:1,nest:1,newlin:1,nhomalt:6,nih:4,nlm:4,nonindexed_vep_cach:12,number:[0,1,5,11],number_of_transcripts_in_gen:1,oad_source_in_memori:11,object:1,onc:1,one:1,onli:[0,1,8,13],onlyid:10,open:9,option:[0,8,10,13],or4f5:1,or4g11p:1,order:13,other:[0,1],otherwis:1,our:1,ourcefil:11,out:[3,8,10,12],out_annot:0,output:[3,8,11],outtyp:[3,8,11],over:[0,3],param:5,paramet:13,park:12,pars:[1,11],part:1,path:[0,5,7,10],pathogen:[6,8,13],pcaso:[],per:1,percent_of_transcripts_affect:1,pip:[8,9],pipe:[0,1],pleas:10,plit_multi_allelic_vari:11,point:1,polyphen:[6,13],popul:[1,8],posit:11,post:1,pre:11,precal:[11,12],precalvep:12,pred:[6,13],predict:1,preprocess:[7,11],prerequisit:8,present:1,previou:11,print:11,process:1,program:11,protein_cod:1,provid:8,pyfaidx:9,pyliftov:[0,9,13],pypi:[8,13],pysam:9,pytabix:9,pytest:13,python:[8,9],pyvcf:9,q8nh21:1,qual:1,qualiti:11,read:0,readabl:1,readili:1,record:1,ref:[0,1],refactor:13,refer:[0,7,12],refver:[],refvers:7,region:8,region_vcf:13,releas:[3,8,12],remov:[5,8,11,13],remove_unannotated_vari:13,repres:1,requir:[8,10],resourc:3,restrict:8,root:8,roughli:1,run:[8,9],run_vep:12,s6ahfq0gdn99uu8:3,sampl:[0,13],samplegeno:[0,13],samplevariantkei:13,score:[6,13],section:1,see:[0,10],semicolon:1,separ:[0,8],seri:8,server:10,set:[0,13],setup:9,sever:0,shell:9,should:[1,10],show:11,shown:1,sift:[6,13],simpl:0,sinc:1,singl:[1,5,8,13],single_datasource_fil:10,single_source_mod:[0,3,13],siphy_29way_pi:13,size:[11,13],snpeff:[1,2],snpefflof:1,snv:[1,8],softwar:12,some:[0,5,13],sor:1,soruc:5,sourc:[3,4,8,13],source1:5,source_path:7,sourcefil:[3,5,8,10],sourcefile_path:5,sourfil:0,space:[1,13],spars:11,specif:8,specifi:1,spliceai:[0,3,6],split:[1,6,8,11,13],split_multi_allelic_vari:[3,8,13],strcture:8,string:[1,5],structur:[1,8,11,13],sub:[1,7,11,13],sub_embedded_group:1,subembed:[1,5],subfield:1,summari:8,support:[0,13],tab:1,tabl:[1,11,13],tag:[1,3,8,13],tandem:1,tba:[4,5],tbd:1,temporari:1,test:[3,8],text:0,than:1,them:[0,7,9],thi:[0,1,5,11],those:[0,1],through:1,thu:1,tild:1,titl:11,toolkit:8,total:0,transcribed_unprocessed_pseudogen:1,transcript:[1,13],truncat:1,tsi:[0,3,5,10],turn:[11,13],two:[0,1],type:[1,5,8,11,13],ucsc:0,unannot:13,updat:13,upi0000041bc1:1,upstream_gene_vari:1,usag:[],use:[0,1,9,10,13],used:1,user:13,using:[0,9],v97:1,v99:4,val:1,val_1:1,val_2:1,val_3:1,valid:11,valu:[1,13],variabl:5,variant:[1,5,8,11,13],variant_class:13,variat:4,variou:10,varsom:2,vartyp:[3,10,13],vcf:[3,5,8,10,11,13],vep:[1,2,6,8,12,13],vep_domain:1,vepcach:12,version:[0,1,4,5,7,8,10,11],version_d:[5,10],view:11,visual:1,vqslod:1,vqsrtranchesnp99:1,wai:1,walk:1,want:[0,5,10],web:[11,13],when:[0,13],where:1,which:[0,1,5],whitespac:1,within:1,without:13,work:[],would:1,written:0,www:[3,4],xmltodict:9,you:[0,1,5,9,10]},titles:["<span class=\"section-number\">5. </span>Annotate Variants (<code class=\"docutils literal notranslate\"><span class=\"pre\">annot</span></code>)","<span class=\"section-number\">9. </span>Annotated VCF file format","<span class=\"section-number\">11. </span>Benchmark","<span class=\"section-number\">10. </span>CGAP Example","<span class=\"section-number\">6. </span>Datasource List","<span class=\"section-number\">7. </span>Data Structure File","<span class=\"section-number\">8. </span>Data Structure File Version","<span class=\"section-number\">3. </span>Download Data Source (<code class=\"docutils literal notranslate\"><span class=\"pre\">download</span></code>)","MutAnno","<span class=\"section-number\">1. </span>Installation","<span class=\"section-number\">4. </span>Make Single Source File (<code class=\"docutils literal notranslate\"><span class=\"pre\">makedata</span></code>)","<span class=\"section-number\">2. </span>Optional arguments","Pre-calculate variant annotation","<span class=\"section-number\">12. </span>Version History"],titleterms:{add:0,addit:0,allel:0,annot:[0,1,3,4,10,11,12],argument:11,attribut:5,basic:5,benchmark:2,cadd:4,calcul:12,cgap:3,clinvar:[4,7],code:9,coding_gene_main_chrom:10,content:8,convert:11,coordin:0,data:[0,4,5,6,7,10,11],databas:[4,7],datasourc:4,dbsnp:4,download:7,exampl:[1,3],field:5,file:[0,1,5,6,10,11],format:[1,11],from:9,full:3,gene:[3,6,10],gener:1,genoinfo:0,genom:10,genotyp:0,github:9,gnomad:4,hg19:0,histori:13,inform:0,input:[0,10],instal:9,item:1,list:4,make:[10,11],makedata:[10,11],micro:3,multi:0,mutanno:8,onli:10,option:11,out:0,output:[0,10],outtyp:0,pathogen:4,popul:4,pre:12,prerequisit:9,pypi:9,region:10,releas:13,remov:0,requir:1,restrict:1,root:5,run:11,separ:1,seri:13,singl:[0,10,11],snv:10,sourc:[0,5,7,10,11],sourcefil:0,specif:[1,10],split:0,split_multi_allelic_vari:0,strcture:5,structur:[0,5,6,10],summari:1,tag:0,type:[0,10],usag:[],variant:[0,4,6,7,12],vcf:[0,1],vep:4,version:[6,13]}})