Search.setIndex({docnames:["annotate_variants","annotated_vcf","benchmark","contents","datasource_list","datastructure","datastructure_version","download","index","installation","make_sourcefile","options","precalculate_annotation","version_list"],envversion:{"sphinx.domains.c":2,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":3,"sphinx.domains.index":1,"sphinx.domains.javascript":2,"sphinx.domains.math":2,"sphinx.domains.python":2,"sphinx.domains.rst":2,"sphinx.domains.std":1,sphinx:56},filenames:["annotate_variants.rst","annotated_vcf.rst","benchmark.rst","contents.rst","datasource_list.rst","datastructure.rst","datastructure_version.rst","download.rst","index.rst","installation.rst","make_sourcefile.rst","options.rst","precalculate_annotation.rst","version_list.rst"],objects:{},objnames:{},objtypes:{},terms:{"00to99":1,"1000gp":1,"3a14825":1,"3a31276":1,"3a65565":1,"4dn":[],"874e":1,"boolean":5,"case":[0,1],"class":11,"default":[0,5,10,13],"function":[1,5,13],"import":0,"new":1,"true":5,And:0,But:[0,10],For:[0,1],POS:[0,1],The:[0,1,5,10],There:[],These:1,With:[0,10],YES:1,abl:1,abov:1,accept:[],access:1,actual:1,add:[5,6,8,11,13],add_genoinfo:13,added:[7,10],addit:[1,8],after:[1,10],again:1,all:[0,1],allel:[1,8,11],allow:1,alreadi:0,also:1,alt:[0,1],altern:[0,9],among:10,ani:[1,11],annot3tool:12,annot:[5,8,13],annotation_sourc:10,annotation_sourcefil:0,annovar:2,anot:[],append:0,appli:[1,13],applic:[],appropri:1,arg:[],argument:8,ariant_class:11,arrai:1,assign:[0,5,10],assum:1,attribut:8,author:[],automat:[0,5,9],avil:5,bam:[],bamsnap:[],base:[0,8],basic:[8,10],becaus:9,bed:5,been:0,begin:1,being:0,below:1,benchmark:8,between:[],bgm:9,bgzip:0,bin:12,bio:12,blank:[1,13],block:[1,11],blocksiz:10,bold:1,both:1,cache_vers:12,cadd:8,calcul:11,can:[0,1,5,9,10],cannot:[0,1],cast:1,ccds30547:1,chain:[0,11,13],chang:13,chr1:[0,1,10],chrom:[0,1],chromosom:13,chrx:[],clean:[],clean_tag:[0,13],clean_tag_list:[],clinvar:0,clone:9,code:8,coding_gene_main_chrom:13,collaps:1,com:9,comma:[0,1],command:[0,5,8,9,11],comparison:2,configur:[],consid:1,consist:1,contain:10,content:3,convchrom:10,convers:1,convert:[1,8],coordin:[8,11],correspond:[0,1],count:6,creat:1,cse:0,culprit:1,data1:12,data:[1,8,13],data_structur:[0,10],datasourc:8,datastructur:0,date:[1,4,5],dbmi:[9,12],dbnsfp:13,dbnsfptranscript:13,dbsnp:[8,10],dcic:[],debug:[11,13],decai:[],definit:[],delimit:5,delin:1,delta:6,depth:0,desc:5,describ:1,descript:[1,5],detail:[0,1],determin:1,dictionari:1,differ:1,direct:1,directli:1,directori:5,don:[0,9,10],done:1,download:[0,5,8,10],downstream_gene_vari:1,dsjson:[],each:[0,1,10],ebug:11,edit:5,edu:0,effect:1,effici:10,either:1,elasticsearch:1,element:1,embed:1,encapsul:1,encod:13,end:1,enet:11,enoinfo:11,ensembl:12,ensg00000186092:1,ensg00000240361:1,ensp00000334393:1,enst00000335137:1,enst00000492842:1,entri:1,equal:1,etc:1,even:1,exampl:[0,8],excesshet:1,exit:11,expect:1,extend:[],extern:13,extract:0,fals:5,fasta:12,featur:0,field1:5,field:[0,1,8,11,13],file:[8,13],filethi:[],filter:[1,13],first:[1,13],fix:13,follow:[0,1,9],format:[0,5,8,10,13],forth:1,from:[0,6,8,10,13],further:1,g19:11,gene:[8,10,11],gene_id:1,gene_main_chrom:13,gene_nam:1,gener:[0,5,8],genet:[],geno_info:13,genoinfo:8,genom:8,genotyp:[8,11],get:[0,1,13],git:9,github:8,given:1,gnomad:[6,8],gnomadgenom:0,goldenpath:0,grab:1,grch38:0,grch38_full_analysis_set_plus_decoy_hla:12,grch38_latest_dbsnp_al:10,grch38d1:12,group:1,gvs:11,gz_coding_mirna_polymorph:[],gzip:0,hain:11,has:[0,1],have:[0,1,10],header:[1,13],heirach:5,help:11,here:[0,10],hg002:[1,13],hg003:13,hg004:13,hg19:[8,11],hg38:0,hg38tohg19:0,hgdownload:0,hgnc:1,hgv:[11,13],histori:8,hms:12,home:12,homo_sapiens_vep:12,homozyg:6,how:1,howev:0,http:[0,9],ilenc:11,imag:[],implement:13,includ:[0,10],index:[],individu:6,inf:[],info:[0,1,11],inform:[1,8,11],ingle_source_mod:11,input:[8,11],insert:[1,10],instal:[8,13],interfac:8,is_avail:5,is_list:5,is_most_severe_transcript:13,item:8,json:[0,5,10,11],keep:0,keep_equal_str:13,kei:1,ksdfmk:[],last:1,later:1,lean_tag:11,level:5,librari:9,librarri:0,liftov:[0,11],line:[0,1,8],link:[],list:[0,1,5,8,13],load:11,locat:1,locksiz:11,log:11,logfil:[],loss:1,makdata:[],make:[1,8],makedata:[0,8,13],manag:10,map:1,mark:1,maxd:6,maximum:6,mean:[1,5],mediat:[],memori:11,merg:13,merge_vep:12,messag:11,metadata:1,metric:11,minor:13,mirna:[],miss:13,mk446:12,mleac:1,mleaf:1,mod:[],mode:11,modifi:1,modul:13,more:[0,10],multi:[1,8,11],multiallel:13,multipl:[0,1],must:[1,10],mutanno:[0,1,5,7,9,10,11,12],mutanno_value_sect:[],mutlipl:0,muttano:1,na12877_sampl:0,na12878:[],na12878_sampl:0,na12879_sampl:[],name:[0,1,4,5,10],need:[0,9],negative_train_sit:1,nest:1,newlin:1,nhomalt:6,nnot:[],nomad:[],nonindexed_vep_cach:12,nonsens:[],number:[0,1,5,11],number_of_transcripts_in_gen:1,oad_source_in_memori:11,object:1,onc:1,one:1,onli:[0,1,8],onlyid:10,ooo:[],open:9,option:[0,8,10,13],or4f5:1,or4g11p:1,other:[0,1],otherwis:1,our:1,ourcefil:11,out:[8,10,12],out_annot:0,output:[8,11],output_annot:[],outtyp:[8,11],over:0,packag:[],page:[],param:5,park:12,pars:[1,11],part:1,path:[0,5,10],pathogen:[6,8,13],per:1,percent_of_transcripts_affect:1,pillow:[],pip:[8,9],pipe:[0,1],pleas:10,plit_multi_allelic_vari:11,point:1,polymorphic_pseudogen:[],polyphen:[6,13],popul:[1,8],posit:11,post:1,pre:11,precal:[11,12],precalvep:12,pred:[6,13],predict:1,preprocess:11,prerequisit:8,present:1,previou:11,print:11,process:1,program:11,protein_cod:1,provid:8,pyfaidx:9,pyliftov:[0,9,13],pylisfov:[],pypi:[8,13],pysam:9,pytabix:9,pytest:13,python:[8,9],pyvcf:9,q8nh21:1,qual:1,qualiti:11,quick:[],read:0,readabl:1,readili:1,record:1,ref:[0,1],refer:[0,12],region:8,region_vcf:13,releas:[8,12],remov:[5,8,11,13],remove_unannotated_vari:13,repres:1,requir:[8,10],restrict:8,result:[],rmcsq:[],root:8,roughli:1,run:[8,9],run_vep:12,sampl:[0,13],samplegeno:[0,13],score:[6,13],search:[],section:1,see:[0,10],semicolon:1,separ:[0,8],seri:8,server:10,set:0,setup:9,sever:0,shell:9,should:[1,10],show:11,shown:1,sift:[6,13],silenc:[],simpl:0,sinc:1,singl:[1,5,8],single_datasource_fil:10,single_source_mod:0,siphy_29way_pi:13,site:[],size:11,snpeff:[1,2],snpefflof:1,snv:[1,10],softwar:12,some:[0,5,13],sor:1,sort:[],soruc:5,sourc:[4,8],source1:5,sourcefil:[5,8,10],sourcefile2:[],sourcefile_path:5,sourfil:0,space:[1,13],spars:11,specif:8,specifi:1,spliceai:[0,6],split:[1,6,8,11,13],split_multi_allelic_vari:[8,13],start:[],strcture:8,string:[1,5],structur:[1,8,11,13],sub:[1,11],sub_embedded_group:1,subembed:[1,5],subfield:1,summari:8,support:0,tab:1,tabl:[1,11,13],tag:[1,8,13],tandem:1,tba:[4,5],tbd:1,temporari:1,test:8,text:0,than:1,them:[0,9],ther:[],thi:[0,1,5,11],those:[0,1],through:1,thu:1,tibanna:[],tild:1,titl:11,toolkit:8,total:0,transcribed_unprocessed_pseudogen:1,transcript:[1,13],truncat:1,tsi:[0,5,10],turn:11,two:[0,1],txt:[],type:[1,5,8,11,13],ucsc:0,unannot:13,updat:13,upi0000041bc1:1,upstream_gene_vari:1,use:[0,1,9,10],used:1,using:[0,9],v97:1,v99:4,val:1,val_1:1,val_2:1,val_3:1,valid:11,valu:[1,13],variabl:5,variant:[1,5,8,11,13],variou:10,varsom:2,vartyp:[10,13],vcf:[5,8,10,11],vcf_info_valu:[],vep:[1,2,6,8,12,13],vep_domain:1,vep_select_microannot_add_most_sever:[],vepcach:12,version:[0,1,4,5,8,10,11],version_d:[5,10],view:11,visual:1,vqslod:1,vqsrtranchesnp99:1,wai:1,walk:1,want:[0,5,10],web:[11,13],when:0,where:1,which:[0,1,5],whitespac:1,within:1,would:1,written:0,xmltodict:9,you:[0,1,5,9,10],your:[]},titles:["<span class=\"section-number\">5. </span>Annotate Variants (<code class=\"docutils literal notranslate\"><span class=\"pre\">annot</span></code>)","<span class=\"section-number\">9. </span>Annotated VCF file format","<span class=\"section-number\">10. </span>Benchmark","&lt;no title&gt;","<span class=\"section-number\">6. </span>Datasource List","<span class=\"section-number\">7. </span>Data Structure File","<span class=\"section-number\">8. </span>Data Structure File Version","<span class=\"section-number\">3. </span>Download Data Source (<code class=\"docutils literal notranslate\"><span class=\"pre\">download</span></code>)","MutAnno","<span class=\"section-number\">1. </span>Installation","<span class=\"section-number\">4. </span>Make Single Source File (<code class=\"docutils literal notranslate\"><span class=\"pre\">makedata</span></code>)","<span class=\"section-number\">2. </span>Optional arguments","<span class=\"section-number\">3. </span>Pre-calculate variant annotation","<span class=\"section-number\">11. </span>Version History"],titleterms:{add:0,addit:0,allel:0,annot:[0,1,4,10,11,12],argument:11,attribut:5,basic:5,benchmark:2,cadd:4,calcul:12,clean:[],code:9,configur:[],content:8,convert:11,coordin:0,data:[0,4,5,6,7,10,11],datasourc:4,datastructur:[],dbsnp:4,download:7,exampl:1,field:5,file:[0,1,5,6,10,11],format:[1,11],from:9,gene:6,gener:1,genoinfo:0,genom:10,genotyp:0,github:9,gnomad:4,hg19:0,histori:13,indic:[],inform:0,input:[0,10],instal:9,item:1,json:[],list:4,makdata:[],make:[10,11],makedata:[10,11],multi:0,multiallel:[],mutanno:8,onli:10,option:11,our:[],out:0,output:[0,10],outtyp:0,pathogen:4,popul:4,pre:12,prerequisit:9,pypi:9,quick:[],region:10,releas:13,remov:0,requir:1,restrict:1,root:5,run:11,separ:1,seri:13,singl:[0,10,11],sourc:[0,5,7,10,11],sourcefil:0,specif:[1,10],split:0,split_multi_allelic_vari:0,start:[],strcture:5,structur:[0,5,6,10],summari:1,tabl:[],tag:0,type:[0,10],variant:[0,4,6,12],vcf:[0,1],vep:4,version:[6,13]}})